ldap/servers/slapd/daemon.c

```diff
@@ -831,7 +830,7 @@ accept_thread(void *vports)
 
     while (!g_get_shutdown()) {
         /* Do we need to accept new connections? */
-        int accept_new_connections = ((ct->size - g_get_current_conn_count()) > slapdFrontendConfig->reservedescriptors);
+        int accept_new_connections = (ct->size > ct->conn_next_offset);
```

Comment:
IMHO a better way for the accept_thread to know if there is "room" for a new connection is to use the freelist offset. A downside of using the freelist offset is that it is protected by a lock, but  I didn't want to lock and unlock in the accept_thread loop. 

Apart from the accept_thread, the only other user of the freelist offset are the connection table list threads, where they move connections back onto the freelist, decrementing the offset. So in a worse case scenario the offset can be reduced by a max of 1, meaning a single connection slot could be unused, IMHO this is better than adding the overhead of locking/unlocking here. WDYT

Reply:
IIUC I agree with your proposal. However a side effect when connection table is close to be full is that we may erroneously believe the CT is full AND wait for 250ms before accepting a new connection.
In case we "believe" there is no room, (ct->size <= ct->conn_next_offset), we may double check the value of conn_next_offset (holding the lock) before sleeping. Most of the time we will not have to double check, so it will benefit from the fix and the lock payload is only when CT is almost full.